### PR TITLE
feat: create migration mechanism - update ssr

### PR DESCRIPTION
**Pull request by kpawelczak**
closes: [CXSPA-5613](https://jira.tools.sap/browse/CXSPA-5613)
In this PR we are adding a migration mechanism for applications that use Spartacus with SSR.

**updateServerFiles**: Main function for migration script. A function that returns a schematic rule chain to update server files by removing deprecated imports and files and modifying import paths.

**removeServer**: A rule function that removes the server file and renames the backup server file to replace it.

**modifyServerImports**: A rule function that modifies server imports by replacing deprecated import statements with updated ones.

**removeExpressTokensFile**: A rule function that removes the Express tokens file if it exists in the project.

**getExpressImportPaths**: A utility function that retrieves all import paths related to Express from a TypeScript source file.

**updateTokenImportPaths**: A rule function that updates import paths in TypeScript files by replacing deprecated Express-related import paths with updated ones.

### CODE REVIEW

File: projects/schematics/src/migrations/mechanism/update-ssr/update-ssr.collection.json
Comment (pawelfras) [inline projects/schematics/src/migrations/mechanism/update-ssr/update-ssr.collection.json:1]: is this file required?

File: projects/schematics/src/migrations/mechanism/update-ssr/update-ssr-files.ts
Comment (pawelfras) [inline projects/schematics/src/migrations/mechanism/update-ssr/update-ssr-files.ts:1]: The folder `migrations/mechanism` seems to contain mechanisms for autogenerated migrations based on breaking changes. I propose to move this logic to `2211_0` as it is related strictly to the next-major release and probably won't be reused in future migrations.

File: projects/schematics/src/shared/utils/file-utils.ts [hunk 1]
```diff
@@ -1302,5 +1303,9 @@ export function getLineFromTSFile(
 export function getServerTsPath(host: Tree): string | undefined {
   const projectName = getDefaultProjectNameFromWorkspace(host);
   const angularJson = getAngularJsonFile(host);
-  return angularJson.projects[projectName].architect?.server?.options?.main;
+
+  return (
+    angularJson.projects[projectName].architect?.server?.options?.main ||
+    (angularJson.projects[projectName].architect as any)?.build.options.server
+  );
```
Comment (pawelfras) [inline projects/schematics/src/shared/utils/file-utils.ts]: Can you help me understand why is this change required?

File: projects/schematics/src/shared/utils/package-utils.ts [hunk 2]
```diff
@@ -150,9 +150,11 @@ export function checkIfSSRIsUsed(tree: Tree): boolean {
   const angularFileBuffer = buffer.toString(UTF_8);
   const angularJson = JSON.parse(angularFileBuffer);
   const isServerConfiguration =
+    !!angularJson.projects[projectName].architect.build.options.server ||
```
Comment (pawelfras) [inline projects/schematics/src/shared/utils/package-utils.ts]: Can you help me understand why this change is required?

**Review by pawelfras — COMMENTED** [review]
As a part of this PR, let's remove from the [2211-migration.md file](https://github.com/SAP/spartacus/blob/aaf01e1f0344f9554a9b1a22c773190816861579/docs/migration/2211_0/2211-migration.md#L1) all steps that are going to be automatized by these schematics.

File: projects/schematics/src/shared/utils/file-utils.ts (hunk 1 above)
Comment (kpawelczak) [reply projects/schematics/src/shared/utils/file-utils.ts]: This function is part of checkIfSsrIsUsed. Given the changes in the angular.json structure for new Angular 17 apps, adjustments were made to accommodate these differences.

File: projects/schematics/src/shared/utils/package-utils.ts (hunk 2 above)
Comment (kpawelczak) [reply projects/schematics/src/shared/utils/package-utils.ts]: Given the changes in the angular.json structure for new Angular 17 apps, adjustments were made to accommodate these differences. Without this adjustment, the detection of SSR usage was not functioning correctly.

File: projects/schematics/src/migrations/mechanism/update-ssr/update-ssr.collection.json
Comment (kpawelczak) [reply projects/schematics/src/migrations/mechanism/update-ssr/update-ssr.collection.json:1]: yes, it's used for unit-testing.

File: projects/schematics/src/shared/utils/file-utils.ts (hunk 1 above)
Comment (pawelfras) [reply projects/schematics/src/shared/utils/file-utils.ts]: Thank you for the explanation. Do these changes occur also in applications migrated to Angular 17? Or only in fresh Angular 17 apps? I'm asking, because I couldn't find such a property in the `angular.json` file in the migrated application.

If it's only the fresh app case, maybe we don't need to worry about it right now as the `checkIfSsrIsUsed` function is used only during the migration.

File: projects/schematics/src/shared/utils/package-utils.ts (hunk 2 above)
Comment (pawelfras) [reply projects/schematics/src/shared/utils/package-utils.ts]: https://github.com/SAP/spartacus/pull/18430#discussion_r1481188430

File: projects/schematics/src/shared/utils/file-utils.ts (hunk 1 above)
Comment (kpawelczak) [reply projects/schematics/src/shared/utils/file-utils.ts]: yes, you are right. I was running these schematics without migrations and misunderstood it as a bug.

File: projects/schematics/src/migrations/mechanism/update-ssr/update-ssr-files.ts
Comment (kpawelczak) [reply projects/schematics/src/migrations/mechanism/update-ssr/update-ssr-files.ts:1]: I've moved the folder to 2211_0

File: projects/schematics/src/shared/utils/file-utils.ts (hunk 1 above)
Comment (kpawelczak) [reply projects/schematics/src/shared/utils/file-utils.ts]: I've removed this part

File: projects/schematics/src/shared/utils/package-utils.ts (hunk 2 above)
Comment (kpawelczak) [reply projects/schematics/src/shared/utils/package-utils.ts]: I've removed this part

**Review by pawelfras — APPROVED** [review]
QAd and LGTM
